### ISSUE TITLE
Potential conflict between Kafka partitioning and Axon segmentation

### ISSUE THREAD

**Issue by abuijze** [AxonFramework/extension-kafka#1]
@altroy75 commented on [Tue Sep 04 2018](https://github.com/AxonFramework/AxonFramework/issues/762)

After trying for analyze how Kafka topic partitioning will work with Axon event processing segments, I've got an impression that having a Kafka topic with several partitions and at the same time more than one Axon segment can potentially lead to skipped, not processed events.

Consider a situation with a Kafka topic that has two partitions and Axon has two configured Axon processor's segments. Also let's assume two JVMs trying to process events concurrently. In such scenario kafka consumer in each JVM will be assigned with one of the two partitions. Then the event processor in each JVM will try to lock a segment. Now it might happen that an event processor's segment doesn't match some or all events in its kafka consumer's partition (i.e. those events are in the second partition), in which case we will have skipped event messages. (note: Kafka consumers above are part of the same consumer group)

This potential problem stems from the fact that Axon segments mechanism is completely unaware and doesn't take into account Kafka dynamic partition assignments inside a consumer group. Those two work distribution mechanisms (i.e. Axon segments and Kafka partitions) work independently of each other and completely unaware of each other. And this can potentially create a problem.

Thanks,
Alex

---

@ghilainm commented on [Wed Sep 05 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-418623812)

Hi! According to what I understand it does work. The event processor segment is in the end a group of Kafka partitions. Therefore, assigning a segment to an Axon processor in the end results in assigning a group of Kafka partitions. When locking a segment the Axon processor will assign to the Kafka consumer the relevant partitions. Therefore I don't see why we would lose any events.

However, this seems to be as you said two distributions mechanisms and the Axon one seems to be a unnecessary overhead. It could be nice if it could directly rely on the Kafka distribution mechanism.

---

@smcvb commented on [Wed Sep 05 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-418632629)

@nklmish, any insights you might be able to share on this topic?

---

@nklmish commented on [Tue Sep 11 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-420249138)

@altroy75, Thank you for the issue!
> This potential problem stems from the fact that Axon segments mechanism is completely unaware and doesn't take into account Kafka dynamic partition assignments inside a consumer group. 

Agree, I had a discussion with @abuijze regarding this, and one of how we can address this problem is by assigning all partitions to each consumer. For future, we are thinking to switch from tracking processor to subscribing processor. 

---

@altroy75 commented on [Wed Sep 12 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-420677580)

@nklmish, Thanks for following up on this!
By assigning all partitions to each consumer I guess you mean manual assignment (i.e. assign()) and not via the consumer group protocol (i.e. subscribe()), correct?  

---

@abuijze commented on [Fri Oct 19 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-431266593)

A somewhat late reply, but we had an Axon 4.0 release to finish. Fortunately, that's done, and we've had some time to find out what's going on here and discuss a solution.
There is in fact an issue with Kafka partitions and how processors track events. Both Axon and Kafka are trying to take care of which instance receives which events. In a Tracking Processor, Axon needs full control. Right now, if a KafkaTrackingToken doesn't know the position of a specific segment, it will let Kafka take care of it. Since all Processor Threads are running in the same Kafka Consumer Group, this may mean a thread will read from a specific position in the middle (or end) of a Stream. Axon requires that partition to be read from the beginning.

Two things need to be done (if I understood @nklmish recommendations correctly):
- The tracking token should enforce offset 0 if it doesn't have an explicit offset for any partition
- Each thread of a processor should be in a separate consumer group, to prevent Kafka from trying to balance consumers

---

@abuijze commented on [Fri Oct 19 2018](https://github.com/AxonFramework/AxonFramework/issues/762#issuecomment-431266753)

Lastly, since Axon 4, the Kafka codebase has moved to another repository: https://github.com/AxonFramework/extensions-kafka.

**Comment by ghilainm** [comment]
@abuijze Why do you need full control? Can't you do it the other way around and use Kafka to balance partitions across consumers?

Segments are used in Axon in order to implement parallel consumption if I am not wrong. Why can't you assign dynamically partitions to segments based on the Kafka auto balancing of partitions across consumer group?

**Comment by abuijze** [comment]
Hi @ghilainm,
the way tracking processors work, is that the processor itself is in charge of partitioning and ordering. However, I had already discussed with @nklmish that it would be nice to also have a SubscribableMessageSource for Kafka, which allows Kafka to decide which segments are delivered to which nodes. That message source could also support replays by resetting Kafka pointers to the beginning of the stream.

**Comment by gustavomr** [comment]
Any date to have it fixed? Thanks!

**Comment by mamt84** [comment]
Any progress on this?

**Comment by mamt84** [comment]
Any progress on this?

**Comment by mamt84** [comment]
Not sure why this repo was archived but the questions on this link are relevant to the issue discussed here.
https://github.com/AxonFramework/ReferenceGuide/issues/86

**Comment by smcvb** [comment]
Hi @mamt84, let me shed some light on your last question.
The Reference Guide repository you are referencing has been replaced by the [AxonIQ Reference Guide](https://github.com/AxonIQ/reference-guide) repository. Thus, any adjustments or additions you'd like to see introduced in the guide, should be directed towards this new repository.

Now, for the following comment:
> Any progress on this?

At the moment, AxonIQ as main holder of the Axon Framework repositories, has quite a full task list of components to deal with. Things like the main [framework](https://github.com/AxonFramework/AxonFramework/), any of the currently 6 extension, and of course [Axon Server](https://axoniq.io/product-overview/axon-server) as a dedicated message routing (and open source) solution to getting for example your events across.  Although we very much value the Kafka Extension and we will keep on maintaining it, it is one of many components we are working on.

Additionally, I'd like to stress this is an open source product.
Thus anybody, including you, are able to contribute and resolve issues like this.

It is very good to see a maintained interest in the Kafka Extension, so thanks for you showing that, @mamt84. 
However repeating the question will not ensure swift resolution on this issue.

As soon as we are working on an issue, the 'in progress' label will be added.
I thus suggest to stay tuned for when that happens.

Hope this clarifies the situation.

**Comment by mamt84** [comment]
Hi @smcvb ,

Thanks for getting back on this! I have to say that I don't know what prompted me to ask the same question twice, honestly, this may have been unintentionally.

I understand you guys have your hands full with other things on the Axon Framework ecosystem, that's reasonable, but at the same time this repo hasn't move for more than 4 months and the question on when or if this feature will be solved has gone unanswered for 7 months since it was first asked by **gustavomr**, so I believe it's only fair to respectfully ask again (only once!) if this will be resolved.

I also see that this extension doesn't have yet a first release (other than the RC), it also doesn't have any feature branch, fork or any active development. Does this mean the Kafka extension development is not a priority for AxonIQ (as main holder of this repo)?

I know AxonIQ is doubling down on the Axon Server (and maybe that's the reason why this and other event transport extensions haven't seen much development as the Axon Server ever since), but unfortunately this is not an option for my organization since the enterprise version of it, which provides clustering, monitoring and other goodies, doesn't seem to be open source.

That being said, it would be great for me and my team to know if Axon-Kafka extension is or not a priority and will see some degree of active development in the foreseeable future, so that way we know if Axon is an alternative for us or we have to look at something else.

Thanks,
Martin

**Comment by smcvb** [comment]
Thanks for providing this feedback @mamt84 and voicing your concern in regards to progress on the Kafka Extension.
Given the requirement you have, I understand the pressure you are experiencing in regards to the status of this repository.

I'll discuss this with the team, but I am pretty sure the outcome will be in favor of resolving this asap.
As soon as work starts on this issue, the labels will be adjusted accordingly. Thus, stay tuned for future developments.

**Comment by qw3rtrun** [comment]
Hi @mamt84 @smcvb 
Thank you very much for follow-upping this topic. Our team is in a similar situation as Martin's. And the Kafka-extension seems to be a most suitable open-sourced option for distributing events with axon.

**Comment by larsf96** [comment]
@abuijze @smcvb Are there any updates on this topic? I'm currently not able to use the replay feature with Kafka

**Comment by smcvb** [comment]
We are currently working hard at finalizing the 4.2 release of Axon Framework and Axon Server which is eating up all our time. 
The plan is to, after this, ensure to get the Kafka Extension into a good state, so that we can provide a 4.0, 4.1 and 4.2 compatible version of it.

I can however not give you a time scale on this, so I can only ask you to stay patient.
Hope you understand @larsf96, sorry to keep you (and all others of course) waiting on this.

**Comment by larsf96** [comment]
@abuijze @smcvb Do you already have an update on this?

**Comment by smcvb** [comment]
As you might have noticed if you follow AxonIQ channels, last couple of weeks have been tremendously busy for us due to webinars, trainings and our conference last Friday.

A little insight: I have a ton of pending changes on my local environment to align this extension with the standard we like to achieve. But prior to that I will be giving precedence over #11 as a added feature for resilience (due to merge conflicts with my local changes). After all this, this issue will be resolved.

I'll thus like to extend my following comment: 

> I can however not give you a time scale on this, so I can only ask you to stay patient.
> Hope you understand @larsf96, sorry to keep you (and all others of course) waiting on this.

Even better, a pull request to resolve any issues in the extension would speed up things considerably.

**Comment by smcvb** [comment]
I've taken the effort to traverse this issue thread, the user group and trail on previous discussion regarding this problem to come to a plan of attack to resolve the discrepancy between Kafka's Partitioning and Axon's Segmentation for Tracking Event Processors.

To that end, I have split up this ticket in #17 and #18.
As such, I will be closing this issue as 'Obsolete'.
